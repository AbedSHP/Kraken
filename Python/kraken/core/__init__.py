"""Kraken Core."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "alpha"


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Return:
    String, current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
