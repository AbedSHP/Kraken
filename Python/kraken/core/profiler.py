"""Kraken - objects.profiler module.

Classes:
Profiler - Profiler Object.

"""
import time




class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None

    class ProfilerItem(object):

        def __init__(self, label):
            super(Profiler.ProfilerItem, self).__init__()

            t = time.time()
            self.label = label
            self.start = t
            self.end = t
            self.children = []

        def addChild(self, item):
            self.children.append(item)

        def endProfiling(self):
            self.end = time.time()

    def __init__(self, label='Root'):
        super(Profiler, self).__init__()
        self.reset(label=label)

    def reset(self, label='Root'):
        """Resets the profiler for generating a new report

        Return:
        None

        """
        self.__roots = []
        self.__stack = []

    def push(self, label):
        """Adds a new child to the profiling tree and activates it.

        Return:
        None

        """

        item = Profiler.ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)
        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to the parent item

        Return:
        None

        """
        end = time.time()
        if len(self.__stack) == 0:
            raise Exception("Unable to close bracket. Pop has been called more times than push.") 
             
        self.__stack[-1].endProfiling()
        self.__stack.pop()
        
    def generateReport(self):
        """Returns a json object that encodes the data gathered during profiling.

        Return:
        the json object

        """
        if len(self.__stack) != 0:
            raise Exception("Profiler brackets not closed properly. pop must be called for every call to push. Pop needs to be called another " + str(len(self.__stack)) + " times") 
        report = {
            'tree': None
        }
        functions = {}
        def reportItem(item):
            itemReport  = {
                'label': item.label,
                'duration' : item.end - item.start
            }
            if item.label not in functions:
                functions[item.label] = itemReport['duration']
            else:
                functions[item.label] += itemReport['duration']

            for childItem in item.children:
                if 'children' not in itemReport:
                    itemReport['children'] = []
                itemReport['children'].append(reportItem(childItem))
            return itemReport

        report['tree'] = []
        for rootItem in self.__roots:
            report['tree'].append(reportItem(rootItem))
        return report

    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Return:
        The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
