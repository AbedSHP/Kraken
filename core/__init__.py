"""Kraken Framework."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "beta"


def getVersion():
    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString