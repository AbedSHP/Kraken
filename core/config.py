import json

envConfig = {
                "app": "xsi",
                "namingTemplate": {
                                    "suffixes": {
                                                    "null":"null",
                                                    "control":"ctrl",
                                                    "zero":"zero",
                                                    "srt":"srt",
                                                    "geometry":"geo",
                                                    "curve":"crv",
                                                    "visualCue":"visCue",
                                                    "implicit":"imp",
                                                    "hierarchy":"hrc",
                                                    "upVector":"upV",
                                                    "envelope":"env",
                                                    "group":"grp",
                                                    "parameterIn":"paramIn",
                                                    "parameterOut":"paramOut",
                                                    "kineIn":"kineIn",
                                                    "kineOut":"kineOut",
                                                    "srtIn":"srtIn",
                                                    "srtOut":"srtOut",
                                                    "root":"rt",
                                                    "effector":"eff",
                                                    "bone":"bone",
                                                    "iceHost":"iceHost",
                                                    "pointCloud":"pcloud"
                                                },
                                    "locations": {
                                                    "left":"L",
                                                    "medial":"M",
                                                    "right":"R"
                                                }
                                  }
            }


def setEnvConfigFromFile(filePath):
    """Sets the 'envConfig' from a json file.

    Arguments:
    filePath -- String, path on disk of json config file.

    """

    with file(filePath, "r") as configFile:
        config = json.load(filePath)

    global envConfig
    envConfig = config

    return