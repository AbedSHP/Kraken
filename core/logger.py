import logging
import os
import sys
import types

LOGGER_PREFIX = 'kraken.pyLogger'


class PyLogger(logging.Logger):

    def fatal(self, msg, *args, **kwargs):
        self.log(45, msg, *args, **kwargs)

    def comment(self, msg, *args, **kwargs):
        self.log(5, msg, *args, **kwargs)


class PyHandler(logging.Handler):

    def emit(self, record):

        if isinstance(record.msg, types.StringTypes):
            msg = self.format(record)
        else:
            record.msg = str(record.msg)
            msg = self.format(record)

        print msg


def getLogger(name=None):

    if name is None:
        return logging.getLogger(LOGGER_PREFIX)
    else:
        return logging.getLogger(LOGGER_PREFIX + '.' + name)


if not hasattr(sys, 'PY_LOGGING_CONFIGURED'):
    logging.setLoggerClass(PyLogger)
    logging.addLevelName(45, 'FATAL')
    logging.addLevelName(5, 'COMMENT')
    logger = logging.getLogger()
    pyHandler = PyHandler()
    pyHandler.setLevel(0)
    #fileHandler = logging.FileHandler(os.path.join(os.environ['XSI_USERHOME'], 'xsiScriptLog.txt'))
    #fileHandler.setFormatter(logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s"))
    #fileHandler.setLevel(45)
    logger.addHandler(pyHandler)
    #log.addHandler(fileHandler)
    logger.setLevel(0)
    sys.PY_LOGGING_CONFIGURED = True
    logger = getLogger()