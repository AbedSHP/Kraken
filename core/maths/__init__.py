"""Kraken - math module."""


from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo


PI = 3.141592653589793
DEG_TO_RAD = 0.017453292519943295
RAD_TO_DEG = 57.29577951308232


def Math_radToDeg(val):
    """Converts radians to degrees.

    Arguments:
    val -- Scalar, value to convert to degrees.

    Return:
    Scalar, value in degrees.

    """

    return val * RAD_TO_DEG


def Math_degToRad(val):
    """Converts degrees to radians.

    Arguments:
    val -- Scalar, value to convert to radians.

    Return:
    Scalar, value in radians.

    """

    return val * DEG_TO_RAD