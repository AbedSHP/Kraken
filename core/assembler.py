"""Kraken - Assembler."""

from kraken.core.objects import containers
from kraken.core import *

from collections import OrderedDict
import json


class Assembler(object):
    """Kraken Assembler."""

    def __init__(self):
        super(Assembler, self).__init__()
        self.container = None
        self.definition = {
                        "app":"Kraken",
                        "version":getVersion(),
                        "definition":{}
                      }


    def addContainer(self, container):
        """Adds a container to the Assembler.

        Arguments:
        container -- Container to add to this Assembler.

        Return:
        True if successful.
        """

        if container.name in self.containers.keys():
            raise KeyError("Container is already part of the Assembler: " + container.name)
            return False

        self.containers[container.name] = container
        container.container = self

        return container


    def removeContainer(self, containerName):
        """Removes a container from this Assembler.

        Arguments:
        containerName -- String, name of the container to remove from the Assembler.

        Return:
        True if successful.
        """

        if componentName not in self.components.keys():
            raise KeyError("Container not found in this Assembler: " + componentName)
            return False

        targetContainer = self.components[componentName]
        targetContainer.container = None

        del self.components[componentName]

        return True


    # ================
    # Build Functions
    # ================
    def buildDef(self):
        """Builds the Rig Definition and stores to rigDef attribute.

        Return:
        Dictionary of object data.
        """

        if self.container is None:
            raise AttributeError("Assembler has no container!")

        self.definition["definition"] = self.container.buildDef()

        return self.definition


from kraken.core.objects.containers import *

assembler = Assembler()
container = Container("MyContainer")
assembler.container = container
print json.dumps(assembler.buildDef())