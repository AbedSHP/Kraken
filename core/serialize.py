"""Kraken - encoder module.

Classes:
KrakenEncoder -- Kraken JSON encoder.
KrakenDecoder -- Kraken JSON decoder.
"""

import math
import json

from kraken.core.maths import vec
from kraken.core.maths.rotation import Euler
from kraken.core.maths.rotation import Quat
from kraken.core.maths.matrix import Matrix33
from kraken.core.maths.matrix import Matrix44
from kraken.core.maths import xfo


TYPES = {
            "Vec2": vec.Vec2,
            "Vec3": vec.Vec3,
            "Vec4": vec.Vec4,
            "Euler": Euler,
            "Quat": Quat,
            "Matrix33": Matrix33,
            "Matrix44": Matrix44,
            "Xfo": xfo.Xfo,
        }


class KrakenJSONEncoder(json.JSONEncoder):
    """JSONEncoder class that can encode Kraken object types.

    Custom objects are encoded as JSON object literals (ie, dicts) with
    one key, '__TypeName__' where 'TypeName' is the actual name of the
    type to which the object belongs.  That single key maps to another
    object literal which is just the __dict__ of the object encoded.

    """

    def default(self, obj):

        if hasattr(obj, "jsonEncode"):
            key = '__%s__' % obj.__class__.__name__
            return { key: obj.jsonEncode() }

        return json.JSONEncoder.default(self, obj)