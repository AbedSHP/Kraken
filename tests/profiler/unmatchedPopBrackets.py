from kraken.core.profiler import Profiler
import time
import json

profiler = Profiler()
time.sleep(0.15)
profiler.push('fn1')
time.sleep(0.05)
profiler.push('fn2')
time.sleep(0.01)
profiler.pop()
profiler.generateReport()
