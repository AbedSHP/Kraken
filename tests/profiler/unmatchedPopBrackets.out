Profiler brackets not closed properly. pop must be called for every call to push. Pop needs to be called another 1 times